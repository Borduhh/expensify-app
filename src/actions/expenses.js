import db from '../firebase/firebase';

// ADD_EXPENSE Action
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

// Async start add expense
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0
    } = expenseData;

    return db.ref('expenses').push({ description, note, amount, createdAt })
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          description,
          note,
          amount,
          createdAt
        }));
      });

  };
};

// EDIT_EXPENSE Action
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return(dispatch) => {
    return db.ref(`expenses/${ id }`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  }
};

// REMOVE_EXPENSE Action
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return db.ref(`expenses/${ id }`).remove().then(() => {
      dispatch(removeExpense({id}));
    });
  };
};

// SET_EXPENSES Action
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// Async set expenses
export const startSetExpenses = () => {
  return (dispatch) => {
    return db.ref('expenses').once('value').then((snapshot) => {
        const expenseData = [];

        snapshot.forEach((expense) => {
          expenseData.push({
            id: expense.key,
            ...expense.val()
          });
        });

        dispatch(setExpenses(expenseData));
      });
  };
};