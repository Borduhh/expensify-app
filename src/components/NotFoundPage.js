import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
  <div>
    This is the Not Found Page component.
    <Link to='/'>Homepage</Link>
  </div>
);

export default NotFoundPage;