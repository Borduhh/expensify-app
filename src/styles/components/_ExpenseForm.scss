.expense-form {
  .input-group {
      background:#444;
      
    .input-group__item {
      width:49%;  
    }
  }
}