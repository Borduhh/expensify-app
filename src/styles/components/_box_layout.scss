.box-layout {
  align-items: center;
  background: url('/images/bg.jpg');
  
  background-size: cover;
  display:flex;
  height: 100vh;
  justify-content: center;
  width: 100vw;
}

.box-layout__box {
  background: fade-out($color: #FFFFFF, $amount: 0.15);
  border-radius: 3px;
  padding: $xl-size $l-size;
  max-width: 35rem;
  text-align: center;
}

.box-layout__title {
  line-height: 1;
  margin: 0 0 $m-size 0;
}