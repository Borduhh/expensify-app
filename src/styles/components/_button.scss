.button { 
  background: $light-blue;
  border: solid $mid-blue;
  border-radius: 2px;
  border-width: 0 0 5px;
  color: $white;
  font-size: $fs-large;
  font-weight: 200;
  padding: $m-size;
  text-decoration: none;

  transition: $trans-what-all $trans-time-normal $trans-how-ease;
  -webkit-transition: $trans-what-all $trans-time-normal $trans-how-ease;
  -moz-transition: $trans-what-all $trans-time-normal $trans-how-ease;   
  -ms-transition: $trans-what-all $trans-time-normal $trans-how-ease;   
  -o-transition: $trans-what-all $trans-time-normal $trans-how-ease;

  &:hover {
    background: $mid-blue;
    border-color: $dark-blue;
  }
}

.button--thin {
  border-width: 0 0 3px;
  font-size: $fs-medium;
  padding: $xs-size $l-size;
}

.button--link {
  background:none;
  border:none;
  padding: 0;
}