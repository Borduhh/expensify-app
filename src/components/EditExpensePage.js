import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';

import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  handleOnSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  handleOnClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h2 className='page-header__title'>Edit Expense</h2>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.handleOnSubmit} />
          <button className='button--secondary-full-width' onClick={this.handleOnClick}>Remove Expense</button>
        </div>
      </div>
    );
  };
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    startEditExpense: (id, expense) => { dispatch(startEditExpense(id, expense)) },
    startRemoveExpense: (expenseId) => { dispatch(startRemoveExpense(expenseId)) }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);