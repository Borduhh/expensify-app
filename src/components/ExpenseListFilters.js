import React from 'react';
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {

  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortByChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  }

  render() {
    return (
      <div className='input-group content-container'>
        <div className='input-group__item input-group__item--text-filter'>
          <input
            type='text'
            className='text-input'
            placeholder='Search Expenses...'
            value={
              this.props.filters.text
            }
            onChange={this.onTextChange}
          />
        </div>
        <div className='input-group__item input-group__item--sort-filter'>
          <select
            className='select'
            value={
              this.props.filters.sortBy
            }
            onChange={this.onSortByChange}
          >
            <option value='date'> Date </option>
            <option value='amount' > Amount </option>
          </select>
        </div>
        <div className='input-group__item input-group__item--date-filter'>
          <DateRangePicker
            startDate={this.props.filters.startDate}
            startDateId='filterDatesStartDate'
            endDate={this.props.filters.endDate}
            endDateId='filterDatesEndDate'
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            showClearDates={true}
          />
        </div>
      </div>
    );
  };
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);