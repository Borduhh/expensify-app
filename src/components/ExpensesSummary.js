import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

import { getVisibleExpenses } from '../selectors/expenses';
import { getExpensesTotal } from '../selectors/expenses-total';

export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
  const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

  return (
    <div className='page-header'>
      <div className='page-header__content content-container'>
        <h1 className='page-header__title'>Viewing <span>{expensesCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span></h1>
        <Link className='button button--thin' to='/create'>Add Expense</Link>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expensesCount: getVisibleExpenses(state.expenses, state.filters).length,
    expensesTotal: getExpensesTotal(getVisibleExpenses(state.expenses, state.filters))
  };
};

export default connect(mapStateToProps)(ExpensesSummary);