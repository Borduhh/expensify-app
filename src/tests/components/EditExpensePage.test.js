import React from 'react';
import { shallow } from 'enzyme';

import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpenseSpy, removeExpenseSpy, historySpy, wrapper;

beforeEach(()=> {
  editExpenseSpy = jest.fn();
  removeExpenseSpy = jest.fn();
  historySpy = {
    push: jest.fn()
  };

  wrapper = shallow(<EditExpensePage 
    editExpense={editExpenseSpy}
    removeExpense={removeExpenseSpy}
    history={historySpy}
    expense={expenses[1]}
  />);
});

// EditExpensePage Render Test
test('Should render EditExpensePage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should handle editExpense onSubmit',() => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);

  expect(historySpy.push).toHaveBeenLastCalledWith('/');
  expect(editExpenseSpy).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('Should handle removeExpense onClick',() => {
  wrapper.find('button').simulate('click');

  expect(historySpy.push).toHaveBeenLastCalledWith('/');
  expect(removeExpenseSpy).toHaveBeenLastCalledWith({ id: expenses[1].id });
});
