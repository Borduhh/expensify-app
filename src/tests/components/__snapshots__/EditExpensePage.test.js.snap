// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage 1`] = `
<div>
  <h2>
    Edit Expense
  </h2>
  <ExpenseForm
    expense={
      Object {
        "amount": 54000,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Expense
  </button>
</div>
`;
